File: C#/Projects/SnomedQuery.Parser/RF2Relationship.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SnomedQuery.Parser
{
  /// <summary>
  /// Stored data from one line of the RFS concept file.
  /// </summary>
  public class RF2Relationship : RF2ItemSingle
  {
    /// <summary>
    /// Snomed Id of relationship source
    /// </summary>
    public Int64 SourceId;

    /// <summary>
    /// Snomed Id of relationship destination
    /// </summary>
    public Int64 DestinationId;

    /// <summary>
    /// Snomed relationship group.
    /// </summary>
    public Int32 RelationshipGroup;

    /// <summary>
    /// Snomed relationship type.
    /// </summary>
    public Int64 TypeId;

    /// <summary>
    /// Snomed relationship characteristic type id.
    /// </summary>
    public Int64 CharacteristicTypeId;

    /// <summary>
    /// Snomed modifier id.
    /// </summary>
    public Int64 ModifierId;

    /// <summary>
    /// Lazy load TypeConcept.
    /// </summary>
    public RF2ConceptGroup TypeConcept
    {
      get
      {
        if (this.typeConcept == null)
          this.typeConcept = this.Parser.GetConceptGroup(this.TypeId);
        return this.typeConcept;
      }
    }

    /// <summary>
    /// Backing field for TypeConcept property.
    /// </summary>
    RF2ConceptGroup typeConcept = null;

    /// <summary>
    /// Lazy load Source.
    /// </summary>
    public RF2ConceptGroup Source
    {
      get
      {
        if (this.source == null)
          this.source = this.Parser.GetConceptGroup(this.SourceId);
        return this.source;
      }
    }

    /// <summary>
    /// Backing field for Source property.
    /// </summary>
    RF2ConceptGroup source = null;

    /// <summary>
    /// Lazy load Destination.
    /// </summary>
    public RF2ConceptGroup Destination
    {
      get
      {
        if (this.destination == null)
          this.destination = this.Parser.GetConceptGroup(this.DestinationId);
        return this.destination;
      }
    }

    /// <summary>
    /// Backing field for Destination property.
    /// </summary>
    RF2ConceptGroup destination = null;

    /// <summary>
    /// Constructor
    /// </summary>
    public RF2Relationship()
    {
    }

    /// <summary>
    /// Parse single line from relationship rf2 file.
    /// </summary>
    /// <param name="parser"></param>
    /// <param name="line"></param>
    /// <returns></returns>
    public static RF2Relationship Parse(RF2Parser parser, String line)
    {
      String[] parts = line.Split('\t');
      if (parts.Length != 10)
        throw new ApplicationException("Invalid Relationship line");

      return new RF2Relationship()
      {
        Parser = parser,
        Id = Int64.Parse(parts[0]),
        EffectiveTime = ParseEffectiveTime(parts[1]),
        Active = ParseBool(parts[2]),
        ModuleId = Int64.Parse(parts[3]),
        SourceId = Int64.Parse(parts[4]),
        DestinationId = Int64.Parse(parts[5]),
        RelationshipGroup = Int32.Parse(parts[6]),
        TypeId = Int64.Parse(parts[7]),
        CharacteristicTypeId = Int64.Parse(parts[8]),
        ModifierId = Int64.Parse(parts[9])
      };
    }

    /// <summary>
    /// Override ToString method.
    /// </summary>
    /// <returns></returns>
    public override string ToString()
    {
      return $"RF2Relationship [{Id}] Active: {Active} TypeId:{TypeId} Source:{SourceId} Destination:{DestinationId}";
    }
  }
}

File: Java/SnomedQuery/Parser/RF2Relationship.java
package Parser;

/**
 * <p>
 * Stored data from one line of the RFS concept file.
 * </p>
 * 
 * @author Travis Lukach
 */
public class RF2Relationship extends RF2ItemSingle {
	/**
	 * <p>
	 * Snomed Id of relationship source
	 * </p>
	 */
	private long sourceId;

	/**
	 * <p>
	 * Snomed Id of relationship destination
	 * </p>
	 */
	private long destinationId;

	/**
	 * <p>
	 * Snomed relationship group.
	 * </p>
	 */
	private int relationshipGroup;

	/**
	 * <p>
	 * Snomed relationship type.
	 * </p>
	 */
	private long typeId;

	/**
	 * <p>
	 * Snomed relationship characteristic type id.
	 * </p>
	 */
	private long characteristicTypeId;

	/**
	 * <p>
	 * Snomed modifier id.
	 * </p>
	 */
	private long modifierId;

	/**
	 * <p>
	 * Gets <i>sourceId</i>.
	 * </p>
	 * 
	 * @return
	 */
	public long getSourceId() {
		return sourceId;
	}

	/**
	 * <p>
	 * Sets <i>sourceId</i>.
	 * </p>
	 * 
	 * @param sourceId
	 */
	public void setSourceId(long sourceId) {
		this.sourceId = sourceId;
	}

	/**
	 * <p>
	 * Gets <i>destinationId</i>.
	 * </p>
	 * 
	 * @return
	 */
	public long getDestinationId() {
		return destinationId;
	}

	/**
	 * <p>
	 * Sets <i>destinationId</i>.
	 * </p>
	 * 
	 * @param destinationId
	 */
	public void setDestinationId(long destinationId) {
		this.destinationId = destinationId;
	}

	/**
	 * <p>
	 * Gets <i>relationshipGroup</i>.
	 * </p>
	 * 
	 * @return
	 */
	public int getRelationshipGroup() {
		return relationshipGroup;
	}

	/**
	 * <p>
	 * Sets <i>relationshipGroup</i>.
	 * </p>
	 * 
	 * @param relationshipGroup
	 */
	public void setRelationshipGroup(int relationshipGroup) {
		this.relationshipGroup = relationshipGroup;
	}

	/**
	 * <p>
	 * Gets <i>typeId</i>.
	 * </p>
	 * 
	 * @return
	 */
	public long getTypeId() {
		return typeId;
	}

	/**
	 * <p>
	 * Sets <i>typeId</i>.
	 * </p>
	 * 
	 * @param typeId
	 */
	public void setTypeId(long typeId) {
		this.typeId = typeId;
	}

	/**
	 * <p>
	 * Gets <i>characteristicTypeId</i>.
	 * </p>
	 * 
	 * @return
	 */
	public long getCharacteristicTypeId() {
		return characteristicTypeId;
	}

	/**
	 * <p>
	 * Sets <i>characteristicTypeId</i>.
	 * </p>
	 * 
	 * @param characteristicTypeId
	 */
	public void setCharacteristicTypeId(long characteristicTypeId) {
		this.characteristicTypeId = characteristicTypeId;
	}

	/**
	 * <p>
	 * Gets <i>modifierId</i>.
	 * </p>
	 * 
	 * @return
	 */
	public long getModifierId() {
		return modifierId;
	}

	/**
	 * <p>
	 * Sets <i>modifierId</i>.
	 * </p>
	 * 
	 * @param modifierId
	 */
	public void setModifierId(long modifierId) {
		this.modifierId = modifierId;
	}

	/**
	 * <p>
	 * Lazy load TypeConcept.
	 * </p>
	 * 
	 * @return
	 */
	public final RF2ConceptGroup getTypeConcept() {
		if (this.typeConcept == null) {
			this.typeConcept = this.getParser().getConceptGroup(this.typeId);
		}
		return this.typeConcept;
	}

	/**
	 * <p>
	 * Backing field for TypeConcept property.
	 * </p>
	 */
	private RF2ConceptGroup typeConcept = null;

	/**
	 * <p>
	 * Lazy load Source.
	 * </p>
	 * 
	 * @return
	 */
	public final RF2ConceptGroup getSource() {
		if (this.source == null) {
			this.source = this.getParser().getConceptGroup(this.sourceId);
		}
		return this.source;
	}

	/**
	 * <p>
	 * Backing field for Source property.
	 * </p>
	 */
	private RF2ConceptGroup source = null;

	/**
	 * <p>
	 * Lazy load Destination.
	 * </p>
	 * 
	 * @return
	 */
	public final RF2ConceptGroup getDestination() {
		if (this.destination == null) {
			this.destination = this.getParser()
					.getConceptGroup(this.destinationId);
		}
		return this.destination;
	}

	/**
	 * <p>
	 * Backing field for Destination property.
	 * </p>
	 */
	public RF2ConceptGroup destination = null;

	/**
	 * <p>
	 * Constructor
	 * </p>
	 */
	public RF2Relationship() {
	}

	/**
	 * <p>
	 * Parse single line from relationship rf2 file.
	 * </p>
	 * 
	 * @param parser
	 * @param line
	 * @return
	 */
	public static RF2Relationship Parse(RF2Parser parser, String line) {
		String[] parts = line.split("[\\t]", -1);
		if (parts.length != 10) {
			throw new RuntimeException("Invalid Relationship line");
		}

		RF2Relationship tempVar = new RF2Relationship();
		tempVar.setParser(parser);
		tempVar.setId(Long.parseLong(parts[0]));
		tempVar.setEffectiveTime(parseEffectiveTime(parts[1]));
		tempVar.setActive(parseBool(parts[2]));
		tempVar.setModuleId(Long.parseLong(parts[3]));
		tempVar.sourceId = Long.parseLong(parts[4]);
		tempVar.destinationId = Long.parseLong(parts[5]);
		tempVar.setRelationshipGroup(Integer.parseInt(parts[6]));
		tempVar.typeId = Long.parseLong(parts[7]);
		tempVar.setCharacteristicTypeId(Long.parseLong(parts[8]));
		tempVar.setModifierId(Long.parseLong(parts[9]));
		return tempVar;
	}

	/**
	 * <p>
	 * Override ToString method.
	 * </p>
	 * 
	 * @return
	 */
	@Override
	public String toString() {
		return String.format(
				"RF2Relationship [%1$s] Active: %2$s TypeId:%3$s Source:%4$s Destination:%5$s",
				getId(), isActive(), typeId, sourceId, destinationId);
	}

}